### PR TITLE
Replace FlowRef with Arc<Flow>, now that Arc supports DST.

### PR DESCRIPTION
**Pull request by SimonSapin**
… and `WeakFlowRef` with `Weak<Flow>`.

r? @pcwalton

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7237)

### CODE REVIEW

**Comment by SimonSapin** [comment]
@pcwalton In `components/layout/parallel.rs`, what makes it safe to to transmute to `UnsafeFlow` and back? Won’t ever flows be destroyed by the time they’re used?

**Comment by pcwalton** [comment]
I'm uncomfortable with the amount of unsafe code this introduces throughout layout. The whole idea of parallel layout is that you are allowed to mutably access the flow you're working on *only* (as well as its children, but not its siblings). That's why I (tried to) forbid `FlowRef`s from leaking into the guts of layout.

One way to do this would be to encapsulate the `Arc`-to-`FlowRef` transmutation into the traversal module and have it hand an `&mut Flow` to the kernel functions. Later on, when we want to eliminate the undefined behavior, we can switch `Flow` to be an `UnsafeCell`.

**Comment by SimonSapin** [comment]
> I'm uncomfortable with the amount of unsafe code this introduces throughout layout

`DerefMut for FlowRef` was unsound, according to #6503. My understanding is that this PR introduces more `unsafe` blocks, but only for code that was already unsafe. I could make the `flow_ref::deref_mut` function non-`unsafe`, but it would only be sweeping #6503 back under the metaphorical carpet.

> The whole idea of parallel layout is that you are allowed to mutably access the flow you're working on only

How can we make sure this rule is not violated in the current code or when reviewing layout code changes?

> `Arc`-to-`FlowRef` transmutation

`FlowRef` is `Arc<Flow>` in this PR. Do you mean unsafe deref of `Arc<Flow>` to `&mut Flow`?

> Later on, when we want to eliminate the undefined behavior, we can switch Flow to be an UnsafeCell.

I don’t understand how `UnsafeCell` would be used, can you explain some more?

**Comment by pcwalton** [comment]
> DerefMut for FlowRef was unsound, according to #6503.

It's only unsound in that it relies on undefined behavior (that the Rust compiler doesn't currently exploit, and likely won't for a long time). It's not actually unsafe, because FlowRefs never leak to the guts of layout. This is something that was designed a couple of years ago; the comments in #6503 are missing all the background.

> How can we make sure this rule is not violated in the current code or when reviewing layout code changes?

One way would be to encapsulate the transmutation from `FlowRef` to `&mut Flow` into the traversal function and to have every layout function receive an `&mut Flow` as an argument. That way, the unsafe transmutation is only in one place, and additional layout code never has to grow more unsafe code.

**Comment by SimonSapin** [comment]
Getting `&mut T` out of `Arc<T>` with more than one reference is unsafe if any other reference could be dereferenced at the same time. If we happen to never do that we’re good, but I’d like to have more assurances that this is the case.

In the mean time, at least having `flow_ref::deref_mut` be a function (instead of a `DerefMut` impl that can be called implicitly) make this more visible. I’ve removed its `unsafe` marker. How does this look?

**Comment by pcwalton** [comment]
The tree structure should ensure that it's safe. However, that crucially depends on the guts of layout never being able to see the `FlowRef`s of siblings and parents (and absolute containing block links). The current code is (or at least was) carefully crafted to try to avoid this (hence all the comments around `ContainingBlockLink`).

**Comment by SimonSapin** [comment]
Ok. So how does this PR look now?

**Comment by SimonSapin** [comment]
Rebased.

**Comment by SimonSapin** [comment]
Rebased.

File: components/layout/flow_ref.rs [hunk 1]
```diff
@@ -8,194 +8,19 @@
...
-            }
-        }
-    }
-}
-
-fn base<'a>(r: &WeakFlowRef) -> &'a BaseFlow {
-    let data = r.object.data;
-    debug_assert!(!data.is_null());
+// FIXME(https://github.com/servo/servo/issues/6503) This introduces unsound mutable aliasing.
+// Try to replace it with Arc::get_mut (which checks that the reference count is 1).
```
Comment (pcwalton) [inline components/layout/flow_ref.rs]: I don't think you will succeed. The reference count will pretty much never be 1, since it's a doubly-linked tree.

**Comment by pcwalton** [comment]
This looks OK except for the goal in the comment. I think that we should instead fix the undefined behavior problem in the long term (with `UnsafeCell`) and prevent `FlowRef` from leaking to the rest of layout. One easy way we could do this is to use privacy to prevent `deref_mut()` from being called outside of certain modules such as the traversal module. At a glance, that looks like it leaves `construct.rs` and the various fragments that contain flows. We could conceivably make `restyle_damage` a Cell to eliminate a whole bunch of `deref_mut`s. For the fragments that contain flows, the flows should always be immediate children of the parent flow, so we could just assert that before handing out the `deref_mut`. (To be specific, we would have a public `fn deref_child_flow_mut(parent: &mut Flow, child: &FlowRef) -> &mut Flow` that asserts that `child` is an immediate child of `parent` before handing back the `&mut Flow`.)

**Comment by SimonSapin** [comment]
I’ve rephrased the comment.

I don’t understand what you mean with `UnsafeCell`. It’s not gonna help if two threads try to mutate the same flow at the same time. And cells can’t (usefully) be used inside `Arc` as they’re `!Sync`.

**Comment by pcwalton** [comment]
Two threads can't mutate the same flow at the same time due to the tree structure.

**Comment by SimonSapin** [comment]
@bors-servo r=pcwalton

> &lt;SimonSapin>	pcwalton: so, can we land #7237 now?
> &lt;pcwalton>	yeah, it looks OK to me

**Pull request merged**